oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,402 @@
+
+Abstract
+========
+
+Proposes best practices for declaring and maintaining dependencies on other
+Python packages in Open edX software repositories.
+
+Motivation
+==========
+
+The Open edX project includes dozens of Python software repositories, most of
+which depend on certain other Python packages being installed in order to
+function correctly.  The simple methods we originally used to do this have
+assorted drawbacks that have repeatedly caused problems over the past few
+years: accidental upgrades to incompatible versions, strict installation
+requirements that restrict the ability of downstream packages to manage their
+own dependency versions, lack of clarity regarding the full set of packages
+actually depended upon, etc.
+
+Outlined here is a recommended standard for declaring dependencies on other
+Python packages which resolves most of these issues and will let us make all
+the Open edX Python packages consistent with each other (and many other open
+source Python projects) for ease of understanding and maintenance.
+
+Specification
+=============
+
+The key to successful Python dependency management in Open edX repositories
+is to break it down into five parts:
+
+1. Identify the different contexts in which dependencies will need to be
+   installed.
+2. For each of these contexts, declare the direct dependencies that will be
+   needed.  Use the least restrictive constraints which should allow pip to
+   install a working set of dependencies.  We generally no longer want or need
+   to manually pin specific versions; ``make upgrade`` will do this for us.
+3. Run ``make upgrade`` to generate from the high-level dependencies
+   declarations a separate set of requirements files which specify the exact
+   package versions that are known to work for a Python `virtualenv`_ created
+   for that context.
+4. Automate updates of the detailed requirements files for each context.
+5. When making deliberate changes to the repository's dependencies, only edit
+   the files given to ``*.in`` files in the ``requirements`` directory, never
+   the generated ``*.txt`` files.  Both types of files must be committed when
+   changed.
+
+.. _virtualenv: https://virtualenv.pypa.io/
+
+Identify Usage Contexts
+-----------------------
+
+The dependencies of Python software are typically installed or run in a
+variety of different contexts over the course of developing and using it.
+The set of dependencies needed to perform a task can easily vary between these
+contexts.  The dependencies for each context will be captured in a
+separate file in the ``requirements`` directory.  Here are some common
+contexts and the file names often used for them:
+
+* Just the standard set of core dependencies for execution on a production
+  server to perform its primary purpose (``base.in``)
+* Additional dependencies which are only needed when optional extra features
+  of a package are desired
```

Comment:
This is the only bullet point that doesn't have a filename attached. Is that an error?

Reply:
The filename would typically be based on the name of the extra in setup.py, I'll a note to that effect.